File: skellig-junit-runner/src/test/java/org/skellig/runner/config/SimpleMessageTestStepProcessor.java
package org.skellig.runner.config;

import org.skellig.teststep.processing.processor.BaseTestStepProcessor;
import org.skellig.teststep.processing.processor.TestStepProcessor;
import org.skellig.teststep.processing.state.TestScenarioState;
import org.skellig.teststep.processing.validation.TestStepResultValidator;

import java.util.HashMap;
import java.util.Map;
import java.util.Objects;

import static org.skellig.task.TaskUtils.runTask;

public class SimpleMessageTestStepProcessor extends BaseTestStepProcessor<SimpleMessageTestStep> {

    private Map<Object, Object> latestReceivedMessage;

    protected SimpleMessageTestStepProcessor(TestScenarioState testScenarioState,
                                             TestStepResultValidator validator) {
        super(testScenarioState, validator);
    }

    @Override
    protected Object processTestStep(SimpleMessageTestStep testStep) {
        if (testStep.getReceiveFrom() != null) {
             Map<Object, Object> response =
                    runTask(() -> latestReceivedMessage, Objects::nonNull, 500, 3000);
            response.put("receivedFrom", testStep.getReceiveFrom());
            return response;
        } else {
            latestReceivedMessage = createResponse(testStep);
            return latestReceivedMessage;
        }
    }

    private Map<Object, Object> createResponse(SimpleMessageTestStep testStep) {
        Map<Object, Object> response = new HashMap<>();
        response.put("originalRequest", testStep.getTestData());
        response.put("receivedBy", testStep.getReceiver());
        response.put("status", "success");
        return response;
    }

    @Override
    public Class<SimpleMessageTestStep> getTestStepClass() {
        return SimpleMessageTestStep.class;
    }


    public static class Builder {
        private TestScenarioState testScenarioState;
        private TestStepResultValidator validator;

        public Builder withTestScenarioState(TestScenarioState testScenarioState) {
            this.testScenarioState = testScenarioState;
            return this;
        }

        public Builder withValidator(TestStepResultValidator validator) {
            this.validator = validator;
            return this;
        }

        public TestStepProcessor<SimpleMessageTestStep> build() {
            return new SimpleMessageTestStepProcessor(testScenarioState, validator);
        }
    }
}

File: skellig-junit-runner/src/test/kotlin/org/skellig/runner/config/SimpleMessageTestStepProcessor.kt
package org.skellig.runner.config

import org.skellig.teststep.processing.processor.BaseTestStepProcessor
import org.skellig.teststep.processing.processor.TestStepProcessor
import org.skellig.teststep.processing.processor.config.ConfiguredTestStepProcessorDetails
import org.skellig.teststep.processing.processor.config.TestStepProcessorConfig
import org.skellig.teststep.processing.processor.config.TestStepProcessorConfigDetails
import org.skellig.teststep.processing.state.TestScenarioState
import org.skellig.teststep.processing.util.TaskUtils.runTask
import java.util.*

class SimpleMessageTestStepProcessor private constructor(testScenarioState: TestScenarioState?) : BaseTestStepProcessor<SimpleMessageTestStep>(testScenarioState!!) {

    private var latestReceivedMessage: MutableMap<Any, Any?>? = null

    override fun processTestStep(testStep: SimpleMessageTestStep): Any? {
        return if (testStep.receiveFrom != null) {
            val response = runTask({ latestReceivedMessage }, { Objects.nonNull(it) }, 500, 3000)
            response!!["receivedFrom"] = testStep.receiveFrom
            response
        } else {
            latestReceivedMessage = createResponse(testStep)
            latestReceivedMessage
        }
    }

    private fun createResponse(testStep: SimpleMessageTestStep): MutableMap<Any, Any?> {
        return hashMapOf(
            Pair("originalRequest", testStep.testData),
            Pair("receivedBy", testStep.receiver),
            Pair("status", "success")
        )
    }

    override fun getTestStepClass(): Class<SimpleMessageTestStep> {
        return SimpleMessageTestStep::class.java
    }

    class Builder {
        private var testScenarioState: TestScenarioState? = null

        fun withTestScenarioState(testScenarioState: TestScenarioState?) = apply {
            this.testScenarioState = testScenarioState
        }

        fun build(): TestStepProcessor<SimpleMessageTestStep> {
            return SimpleMessageTestStepProcessor(testScenarioState)
        }
    }
}

class SimpleMessageTestStepProcessorConfig : TestStepProcessorConfig<SimpleMessageTestStep> {
    override fun config(details: TestStepProcessorConfigDetails): ConfiguredTestStepProcessorDetails<SimpleMessageTestStep> {
        return ConfiguredTestStepProcessorDetails(
            SimpleMessageTestStepProcessor.Builder()
                .withTestScenarioState(details.state)
                .build(),
            SimpleMessageTestStepFactory(
                details.testStepRegistry,
                details.valueExpressionContextFactory,
                "toCustomFormat"
            )
        )
    }
}